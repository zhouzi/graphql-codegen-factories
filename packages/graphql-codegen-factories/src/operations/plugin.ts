import { Kind, FragmentDefinitionNode, concatAST, DocumentNode } from "graphql";
import {
  oldVisit,
  PluginFunction,
  Types,
} from "@graphql-codegen/plugin-helpers";
import {
  FactoriesOperationsVisitor,
  FactoriesOperationsVisitorRawConfig,
} from "./FactoriesOperationsVisitor";

export const plugin: PluginFunction<
  FactoriesOperationsVisitorRawConfig,
  Types.ComplexPluginOutput
> = (schema, documents, config) => {
  const allAst = concatAST(
    documents.map(({ document }) => document as DocumentNode)
  );
  const fragments = allAst.definitions.filter(
    (d) => d.kind === Kind.FRAGMENT_DEFINITION
  ) as FragmentDefinitionNode[];
  const allFragments: FragmentDefinitionNode[] = [
    ...fragments,
    // `externalFragments` config is passed by the near-operation-file preset.
    // It is an array of fragments declared outside of the operation file.
    ...(config.externalFragments || []).map(({ node }) => node),
  ];

  const visitor = new FactoriesOperationsVisitor(schema, allFragments, config);
  const content = (
    oldVisit(allAst, { leave: visitor }) as DocumentNode
  ).definitions
    .filter((definition) => typeof definition === "string")
    .join("\n");

  return {
    prepend: visitor.getImports(),
    content,
  };
};
