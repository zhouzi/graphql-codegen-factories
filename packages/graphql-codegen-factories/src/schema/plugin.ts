import { printSchema, parse } from "graphql";
import {
  oldVisit,
  PluginFunction,
  Types,
} from "@graphql-codegen/plugin-helpers";
import { FactoriesBaseVisitorRawConfig } from "../FactoriesBaseVisitor";
import { FactoriesSchemaVisitor } from "./FactoriesSchemaVisitor";

export const plugin: PluginFunction<
  FactoriesBaseVisitorRawConfig,
  Types.ComplexPluginOutput
> = (schema, documents, config) => {
  const printedSchema = printSchema(schema);
  const astNode = parse(printedSchema);

  const visitor = new FactoriesSchemaVisitor(schema, config);
  const content = oldVisit(astNode, { leave: visitor })
    .definitions.filter(Boolean)
    .join("\n");

  return {
    prepend: visitor.getImports(),
    content,
  };
};
