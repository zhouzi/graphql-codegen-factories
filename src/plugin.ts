import { printSchema, parse } from "graphql";
import {
  oldVisit,
  PluginFunction,
  Types,
} from "@graphql-codegen/plugin-helpers";
import {
  FactoriesVisitor,
  FactoriesVisitorRawConfig,
} from "./FactoriesVisitor";

export const plugin: PluginFunction<
  FactoriesVisitorRawConfig,
  Types.ComplexPluginOutput
> = (schema, documents, config) => {
  const visitor = new FactoriesVisitor(schema, config);
  const printedSchema = printSchema(schema);
  const astNode = parse(printedSchema);
  const content = oldVisit(astNode, { leave: visitor })
    .definitions.filter(Boolean)
    .join("\n");

  return {
    prepend: visitor.config.typesPath
      ? [
          `import * as ${visitor.config.namespacedImportName} from '${visitor.config.typesPath}';\n`,
        ]
      : [],
    content,
  };
};
