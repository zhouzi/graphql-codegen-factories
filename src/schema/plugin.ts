import { printSchema, parse } from "graphql";
import {
  oldVisit,
  PluginFunction,
  Types,
} from "@graphql-codegen/plugin-helpers";
import {
  FactoriesSchemaVisitor,
  FactoriesSchemaVisitorRawConfig,
} from "./FactoriesSchemaVisitor";

export const plugin: PluginFunction<
  FactoriesSchemaVisitorRawConfig,
  Types.ComplexPluginOutput
> = (schema, documents, config) => {
  const visitor = new FactoriesSchemaVisitor(schema, config);
  const printedSchema = printSchema(schema);
  const astNode = parse(printedSchema);
  const content = oldVisit(astNode, { leave: visitor })
    .definitions.filter(Boolean)
    .join("\n");

  return {
    prepend: visitor.getImports(),
    content,
  };
};
