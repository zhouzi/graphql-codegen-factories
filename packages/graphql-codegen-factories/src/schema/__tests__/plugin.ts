import { buildSchema } from "graphql";
import { plugin } from "../plugin";

describe("plugin", () => {
  it("should create factories with built-in types", async () => {
    const schema = buildSchema(`
      type User {
        id: ID!
        organizationId: ID
        email: String!
        name: String
        age: Int!
        followers: Int
        geo: Geo!
        isUser: Boolean!
        isAdmin: Boolean
        status: UserStatus!
        favouriteFruit: Fruit
        posts: [Post!]!
        subscribers: [User!]
      }

      type Geo {
        lat: Float!
        lon: Float
      }

      type Post {
        id: ID!
      }

      enum UserStatus {
        Activated
        Created
      }

      enum Fruit {
        Pineapple
        Mango
      }
    `);

    const output = await plugin(schema, [], {});
    expect(output).toMatchSnapshot();
  });

  it("should use enums as types", async () => {
    const schema = buildSchema(`
      type User {
        status: UserStatus!
      }

      enum UserStatus {
        Activated
        Created
      }
    `);

    const output = await plugin(schema, [], { enumsAsTypes: true });
    expect(output).toMatchSnapshot();
  });

  it("should use the custom scalar defaults", async () => {
    const schema = buildSchema(`
      type User {
        createdAt: Date!
      }

      scalar Date
    `);

    const output = await plugin(schema, [], {
      scalarDefaults: { Date: "new Date()" },
    });
    expect(output).toMatchSnapshot();
  });

  it("should create factories for inputs", async () => {
    const schema = buildSchema(`
      input PostInput {
        id: ID
        title: String!
      }
    `);

    const output = await plugin(schema, [], {});
    expect(output).toMatchSnapshot();
  });

  it("should create factories for Query and Mutation", async () => {
    const schema = buildSchema(`
      type User {
        id: ID!
      }

      type Query {
        users: [User!]!
      }

      type Mutation {
        createUser(id: ID!): User!
      }
    `);

    const output = await plugin(schema, [], {});
    expect(output).toMatchSnapshot();
  });

  it("should customize the factory name", async () => {
    const schema = buildSchema(`
      type User {
        id: ID!
      }
    `);

    const output = await plugin(schema, [], { factoryName: "new{Type}" });
    expect(output).toMatchSnapshot();
  });

  it("should customize the maybe value default", async () => {
    const schema = buildSchema(`
      type Post {
        title: String
      }
      input PostInput {
        title: String
      }
    `);

    const output = await plugin(schema, [], {
      maybeValueDefault: "undefined",
    });
    expect(output).toMatchSnapshot();
  });

  it("should customize the maybe value default and input maybe value default independently", async () => {
    const schema = buildSchema(`
      type Post {
        title: String
      }
      input PostInput {
        title: String
      }
    `);

    const output = await plugin(schema, [], {
      maybeValueDefault: "undefined",
      inputMaybeValueDefault: "null",
    });
    expect(output).toMatchSnapshot();
  });

  it("should customize the input maybe value default", async () => {
    const schema = buildSchema(`
      input PostInput {
        title: String
      }
    `);

    const output = await plugin(schema, [], {
      inputMaybeValueDefault: "undefined",
    });
    expect(output).toMatchSnapshot();
  });

  it("should support enums with an underscore", async () => {
    const schema = buildSchema(`
      enum UserRole {
        SUPER_ADMIN
        ADMIN
      }
      type User {
        role: UserRole!
      }
    `);

    const output = await plugin(schema, [], {});
    expect(output).toMatchSnapshot();
  });

  it("should support directives", async () => {
    const schema = buildSchema(`
      directive @test on FIELD_DEFINITION

      type User {
        id: String
      }
    `);

    const output = await plugin(schema, [], {});
    expect(output).toMatchSnapshot();
  });

  it("should import types from other file", async () => {
    const schema = buildSchema(`
      type User {
        id: ID!
      }
    `);

    const output = await plugin(schema, [], {
      typesPath: "./types.ts",
      importTypesNamespace: "SharedTypes",
    });
    expect(output).toMatchSnapshot();
  });

  it("should create factories for unions", async () => {
    const schema = buildSchema(`
      type User {
        firstName: String!
        lastName: String!
      }

      type Droid {
        codeName: String!
      }

      union Humanoid = User | Droid
    `);

    const output = await plugin(schema, [], {});
    expect(output).toMatchSnapshot();
  });

  it("should support interfaces", async () => {
    const schema = buildSchema(/* GraphQL */ `
      interface Node {
        id: ID!
      }
      type User implements Node {
        id: ID!
        username: String!
      }
    `);

    const output = await plugin(schema, [], {});
    expect(output).toMatchSnapshot();
  });
});
