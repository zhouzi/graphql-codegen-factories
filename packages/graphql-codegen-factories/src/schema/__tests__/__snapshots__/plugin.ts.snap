// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugin should add descriptions 1`] = `
Object {
  "content": "/** This is a description with /* characters that need to be escaped *\\\\/. */
export function createPostMock(props: Partial<Post> = {}): Post {
  return {
    __typename: \\"Post\\",
    title: null,
    ...props,
  };
}

/** This is a description for an input. */
export function createPostInputMock(props: Partial<PostInput> = {}): PostInput {
  return {
    title: null,
    ...props,
  };
}
",
  "prepend": Array [],
}
`;

exports[`plugin should create factories for Query and Mutation 1`] = `
Object {
  "content": "export function createUserMock(props: Partial<User> = {}): User {
  return {
    __typename: \\"User\\",
    id: \\"\\",
    ...props,
  };
}

export function createQueryMock(props: Partial<Query> = {}): Query {
  return {
    __typename: \\"Query\\",
    users: [],
    ...props,
  };
}

export function createMutationMock(props: Partial<Mutation> = {}): Mutation {
  return {
    __typename: \\"Mutation\\",
    createUser: createUserMock({}),
    ...props,
  };
}
",
  "prepend": Array [],
}
`;

exports[`plugin should create factories for inputs 1`] = `
Object {
  "content": "export function createPostInputMock(props: Partial<PostInput> = {}): PostInput {
  return {
    id: null,
    title: \\"\\",
    ...props,
  };
}
",
  "prepend": Array [],
}
`;

exports[`plugin should create factories for unions 1`] = `
Object {
  "content": "export function createUserMock(props: Partial<User> = {}): User {
  return {
    __typename: \\"User\\",
    firstName: \\"\\",
    lastName: \\"\\",
    ...props,
  };
}

export function createDroidMock(props: Partial<Droid> = {}): Droid {
  return {
    __typename: \\"Droid\\",
    codeName: \\"\\",
    ...props,
  };
}

export function createHumanoidMock(props: Partial<Humanoid> = {}): Humanoid {
  switch(props.__typename) {
    case \\"User\\":
      return createUserMock(props);
    case \\"Droid\\":
      return createDroidMock(props);
    case undefined:
    default:
      return createHumanoidMock({ __typename: \\"User\\", ...props });
  }
}
",
  "prepend": Array [],
}
`;

exports[`plugin should create factories with built-in types 1`] = `
Object {
  "content": "export function createUserMock(props: Partial<User> = {}): User {
  return {
    __typename: \\"User\\",
    id: \\"\\",
    organizationId: null,
    email: \\"\\",
    name: null,
    age: 0,
    followers: null,
    geo: createGeoMock({}),
    isUser: false,
    isAdmin: null,
    status: UserStatus.Activated,
    favouriteFruit: null,
    posts: [],
    subscribers: null,
    ...props,
  };
}

export function createGeoMock(props: Partial<Geo> = {}): Geo {
  return {
    __typename: \\"Geo\\",
    lat: 0,
    lon: null,
    ...props,
  };
}

export function createPostMock(props: Partial<Post> = {}): Post {
  return {
    __typename: \\"Post\\",
    id: \\"\\",
    ...props,
  };
}
",
  "prepend": Array [],
}
`;

exports[`plugin should customize the factory name 1`] = `
Object {
  "content": "export function newUser(props: Partial<User> = {}): User {
  return {
    __typename: \\"User\\",
    id: \\"\\",
    ...props,
  };
}
",
  "prepend": Array [],
}
`;

exports[`plugin should customize the input maybe value default 1`] = `
Object {
  "content": "export function createPostInputMock(props: Partial<PostInput> = {}): PostInput {
  return {
    title: undefined,
    ...props,
  };
}
",
  "prepend": Array [],
}
`;

exports[`plugin should customize the maybe value default 1`] = `
Object {
  "content": "export function createPostMock(props: Partial<Post> = {}): Post {
  return {
    __typename: \\"Post\\",
    title: undefined,
    ...props,
  };
}

export function createPostInputMock(props: Partial<PostInput> = {}): PostInput {
  return {
    title: undefined,
    ...props,
  };
}
",
  "prepend": Array [],
}
`;

exports[`plugin should customize the maybe value default and input maybe value default independently 1`] = `
Object {
  "content": "export function createPostMock(props: Partial<Post> = {}): Post {
  return {
    __typename: \\"Post\\",
    title: undefined,
    ...props,
  };
}

export function createPostInputMock(props: Partial<PostInput> = {}): PostInput {
  return {
    title: null,
    ...props,
  };
}
",
  "prepend": Array [],
}
`;

exports[`plugin should disable descriptions 1`] = `
Object {
  "content": "export function createPostMock(props: Partial<Post> = {}): Post {
  return {
    __typename: \\"Post\\",
    title: null,
    ...props,
  };
}

export function createPostInputMock(props: Partial<PostInput> = {}): PostInput {
  return {
    title: null,
    ...props,
  };
}
",
  "prepend": Array [],
}
`;

exports[`plugin should import types from other file 1`] = `
Object {
  "content": "export function createUserMock(props: Partial<SharedTypes.User> = {}): SharedTypes.User {
  return {
    __typename: \\"User\\",
    id: \\"\\",
    ...props,
  };
}
",
  "prepend": Array [
    "import * as SharedTypes from './types.ts';
",
  ],
}
`;

exports[`plugin should support directives 1`] = `
Object {
  "content": "export function createUserMock(props: Partial<User> = {}): User {
  return {
    __typename: \\"User\\",
    id: null,
    ...props,
  };
}
",
  "prepend": Array [],
}
`;

exports[`plugin should support enums with an underscore 1`] = `
Object {
  "content": "export function createUserMock(props: Partial<User> = {}): User {
  return {
    __typename: \\"User\\",
    role: UserRole.SuperAdmin,
    ...props,
  };
}
",
  "prepend": Array [],
}
`;

exports[`plugin should support interfaces 1`] = `
Object {
  "content": "export function createUserMock(props: Partial<User> = {}): User {
  return {
    __typename: \\"User\\",
    id: \\"\\",
    username: \\"\\",
    ...props,
  };
}
",
  "prepend": Array [],
}
`;

exports[`plugin should use enums as types 1`] = `
Object {
  "content": "export function createUserMock(props: Partial<User> = {}): User {
  return {
    __typename: \\"User\\",
    status: \\"Activated\\",
    ...props,
  };
}
",
  "prepend": Array [],
}
`;

exports[`plugin should use the custom scalar defaults 1`] = `
Object {
  "content": "export function createUserMock(props: Partial<User> = {}): User {
  return {
    __typename: \\"User\\",
    createdAt: new Date(),
    ...props,
  };
}
",
  "prepend": Array [],
}
`;
